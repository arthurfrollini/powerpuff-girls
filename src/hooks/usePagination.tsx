import React, { createContext, ReactNode, useContext, useState } from "react";

import { Show } from "../infra/models/TvShow";

interface PaginationContextData {
  startIndex: number;
  pages: number;
  currentPage: number;
  setItems: React.Dispatch<React.SetStateAction<Show[]>>;
  setCurrentPage: React.Dispatch<React.SetStateAction<number>>;
  currentItems: Show[];
}

interface PaginationProps {
  children: ReactNode;
}

export const PaginationContext = createContext<PaginationContextData>(
  {} as PaginationContextData
);

export function PaginationProvider({ children }: PaginationProps) {
  const [items, setItems] = useState<Show[]>([]);
  const [itemsPerPage] = useState(10);
  const [currentPage, setCurrentPage] = useState(0);

  const startIndex = currentPage * itemsPerPage;
  const endIndex = startIndex + itemsPerPage;
  const currentItems = items.slice(startIndex, endIndex);

  const pages = Math.ceil(items.length / itemsPerPage);

  return (
    <PaginationContext.Provider
      value={{
        startIndex,
        pages,
        setItems,
        currentItems,
        setCurrentPage,
        currentPage,
      }}
    >
      {children}
    </PaginationContext.Provider>
  );
}

export const usePagination = () => useContext(PaginationContext);
