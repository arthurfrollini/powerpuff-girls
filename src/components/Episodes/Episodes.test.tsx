import { render, screen } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import { PaginationProvider } from "../../hooks/usePagination";

import { Episodes } from ".";

import renderer from "react-test-renderer";

const episodes = [
  {
    id: 1,
    name: "Episode 1",
    summary: "Summary 1",
    image: {
      medium: "Image 1",
    },
  },
  {
    id: 2,
    name: "Episode 2",
    summary: "Summary 2",
    image: {
      medium: "Image 2",
    },
  },
  {
    id: 3,
    name: "Episode 3",
    summary: "Summary 3",
    image: {
      medium: "Image 3",
    },
  },
];

describe("Episodes component", () => {
  it("should render a list of episode", () => {
    render(
      <Router>
        <PaginationProvider>
          <Episodes episodes={episodes} />
        </PaginationProvider>
      </Router>
    );

    expect(screen.getByText("Episode 2")).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const tree = renderer
      .create(
        <Router>
          <PaginationProvider>
            <Episodes episodes={episodes} />
          </PaginationProvider>
        </Router>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
