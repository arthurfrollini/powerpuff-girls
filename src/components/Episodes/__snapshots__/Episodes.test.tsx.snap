// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Episodes component matches snapshot 1`] = `
Array [
  <div
    className="sc-bdvvtL kTbYry"
  />,
  <div
    className="sc-eCImPb dsOhdD"
    data-testid="pagination"
  />,
]
`;
